### ISSUE TITLE
Depth Sampler - can't retrieve depth information from shader

### ISSUE THREAD

**Issue by yoyonel** [cprogrammer1994/ModernGL#242]
**Description**

I try to use depth buffer/texture (renderbuffer) in RTT (Render To Target) mode in a shader (for example for shadow rendering).
The depth buffer and texture seems to be rightly fill, but the sampler fetch (inside the shader rendering) always return 0.0 (null value).
I follow your instruction inside texture.py:
```py
            By default depth textures have ``GL_TEXTURE_COMPARE_MODE`` set to
            ``GL_COMPARE_REF_TO_TEXTURE``, meaning any texture lookup will
            return a depth comparison value.

            If you need to read the actual depth value in shaders, setting
            ``compare_func`` to a blank string will set ``GL_TEXTURE_COMPARE_MODE`` to
            ``GL_NONE`` making you able to read the depth texture as a ``sampler2D``::

                uniform sampler2D depth;
                out vec4 fragColor;
                in vec2 uv;

                void main() {
                    float raw_depth_nonlinear = texture(depth, uv);
                    fragColor = vec4(raw_depth_nonlinear);
                }
```
but ... it don't work ...  (always 0.0 return values)

**Proof of code**

```py
"""

"""
import numpy as np
import moderngl
import moderngl as mgl


def main():
    ctx = mgl.create_standalone_context()
    ctx.enable(mgl.DEPTH_TEST)
    prog = ctx.program(
        vertex_shader='''
            #version 330

            in vec3 in_vert;

            void main() {
                gl_Position = vec4(in_vert, 1.0);
            }
        ''',
        fragment_shader='''
            #version 330

            void main() {
                //gl_FragDepth = 0.1234;
            }
        ''',
    )

    prog_for_depth = ctx.program(
        vertex_shader='''
            #version 330

            in vec2 in_vert;
            out vec2 uv;

            void main() {
                uv = in_vert;
                gl_Position = vec4(in_vert * 2.0 - 1.0, 0.0, 1.0);
            }
        ''',
        # from `texture.py`
        fragment_shader='''
            #version 330

            uniform sampler2D depth;
            out vec4 fragColor;
            in vec2 uv;

            void main() {
                float raw_depth_nonlinear = texture(depth, uv).r;
                fragColor = vec4(raw_depth_nonlinear);
            }
        ''',
    )

    prog_for_depth['depth'].value = 0

    vertices = np.array([
        [-1, 1, 0.],
        [-1, -1, 0.],
        [1, -1, 0.],
    ])
    vbo_triangle = ctx.buffer(vertices.astype('f4').tobytes())
    vao_triangle = ctx.simple_vertex_array(prog, vbo_triangle, 'in_vert')

    size = (4, 4)
    depth_texture = ctx.depth_texture(size, alignment=1)  # implicit -> dtype='f4', components=1
    fbo_depth = ctx.framebuffer(depth_attachment=depth_texture)
    fbo_depth.use()
    fbo_depth.clear(depth=1.0)

    depth = np.frombuffer(depth_texture.read(alignment=1), dtype=np.dtype('f4')).reshape(size[::-1])
    print("From depth texture - should all be 1")
    print(depth)  # should all be 1

    # Now we render a triangle in there
    vao_triangle.render()

    # It should have some different values in the middle
    depth_from_dbo = np.frombuffer(depth_texture.read(alignment=1), dtype=np.dtype('f4')).reshape(size[::-1])
    print("From depth_texture - some stuff should be 1s and 0.5 where the triangle lies.")
    print(depth_from_dbo)  # some stuff should be 1s and 0.5 where the triangle lies.

    # fs: Full Screen
    canvas_fs_quad = np.array([0.0, 0.0, 0.0, 1.0, 1.0, 0.0, 1.0, 1.0]).astype('f4')
    vbo_fs = ctx.buffer(canvas_fs_quad.tobytes())
    vao_fs = ctx.simple_vertex_array(prog_for_depth, vbo_fs, 'in_vert')
    cbo_fs = ctx.renderbuffer(size, components=1, dtype='f4')
    fbo_fs = ctx.framebuffer(color_attachments=[cbo_fs])

    fbo_fs.use()
    fbo_fs.clear()
    depth_texture.compare_func = ''  # Disable depth comparison completely (from `texture.py`)
    depth_texture.use()
    #
    vao_fs.render(moderngl.TRIANGLE_STRIP)

    #
    data = np.frombuffer(fbo_fs.read(components=1, dtype='f4'), dtype=np.dtype('f4')).reshape(size[::-1])
    print("From final fbo - some stuff should be 1s and 0.5 where the triangle lies. (like depth_texture)")
    print(data)


if __name__ == '__main__':
    main()
```

**Expected behavior**

```sh
From depth texture - should all be 1
[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]
From depth_texture - some stuff should be 1s and 0.5 where the triangle lies.
[[0.50000006 0.50000006 0.50000006 1.        ]
 [0.50000006 0.50000006 1.         1.        ]
 [0.50000006 1.         1.         1.        ]
 [1.         1.         1.         1.        ]]
From final fbo - some stuff should be 1s and 0.5 where the triangle lies. (like depth_texture)
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]
```
I was expected to have some result in the last stage (equal to 'From depth_texture')

Here a screenshot from apitrace, to inspect which texture/buffer/sampler are actives at the last step (rendering full screen quad)/the last glDrawArrays...:
![image](https://user-images.githubusercontent.com/7001058/44513759-8cbd1880-a6be-11e8-838c-7007ceedbf1c.png)

Textures and FrameBuffer seems to be coherent.

**Version info**

```
┗ python -m moderngl
moderngl 5.4.2
--------------
vendor: NVIDIA Corporation
renderer: GeForce GTX 1060 6GB/PCIe/SSE2
version: 4.6.0 NVIDIA 390.77
python: 3.6.6 (default, Jun 27 2018, 14:44:17) 
[GCC 8.1.0]
platform: linux
code: 460
```

**Comment by einarf** [comment]
Weird. I can take a look a bit closer later today. I'm reading depth textures just fine using sampler2D when visualizing linearlized depth here : https://github.com/Contraz/demosys-py/tree/master/examples#minecraft

![](https://raw.githubusercontent.com/Contraz/demosys-py/master/examples/images/minecraft.png)

I'm using a sampler to adjust the texture parameters instead, so maybe there is a bug with depth texture? Give the sampler a try and see if that makes a difference.
```py
sampler = self.ctx.sampler(filter=(moderngl.LINEAR, moderngl.LINEAR), compare_func='')
```

**Comment by yoyonel** [comment]
Einarf you're rock !
Using a sampler was the solution.
```py
"""

"""
import numpy as np
import moderngl
import moderngl as mgl


def main():
    ctx = mgl.create_standalone_context()
    ctx.enable(mgl.DEPTH_TEST)
    prog = ctx.program(
        vertex_shader='''
            #version 330

            in vec3 in_vert;

            void main() {
                gl_Position = vec4(in_vert, 1.0);
            }
        ''',
        fragment_shader='''
            #version 330

            void main() {
            }
        ''',
    )

    prog_for_depth = ctx.program(
        vertex_shader='''
            #version 330

            in vec2 in_vert;
            out vec2 uv;

            void main() {
                uv = in_vert;
                gl_Position = vec4(in_vert * 2.0 - 1.0, 0.0, 1.0);
            }
        ''',
        # from `texture.py`
        fragment_shader='''
            #version 330

            uniform sampler2D depth;
            
            in vec2 uv;
            out float fragColor;

            void main() {
                float raw_depth_nonlinear = texture(depth, uv).r;
                fragColor = raw_depth_nonlinear;
            }
        ''',
    )

    vertices = np.array([
        [-1, 1, 0.],
        [-1, -1, 0.],
        [1, -1, 0.],
    ])
    vbo_triangle = ctx.buffer(vertices.astype('f4').tobytes())
    vao_triangle = ctx.simple_vertex_array(prog, vbo_triangle, 'in_vert')

    size = (4, 4)
    depth_texture = ctx.depth_texture(size)  # implicit -> dtype='f4', components=1
    fbo_depth = ctx.framebuffer(depth_attachment=depth_texture)

    fbo_depth.use()
    fbo_depth.clear(depth=1.0)

    depth = np.frombuffer(depth_texture.read(), dtype=np.dtype('f4')).reshape(size[::-1])
    print("From depth texture - should all be 1")
    print(depth)  # should all be 1

    # Now we render a triangle in there
    vao_triangle.render()

    # It should have some different values in the middle
    depth_from_dbo = np.frombuffer(depth_texture.read(), dtype=np.dtype('f4')).reshape(size[::-1])
    print("From depth_texture - some stuff should be 1s and 0.5 where the triangle lies.")
    print(depth_from_dbo)  # some stuff should be 1s and 0.5 where the triangle lies.

    # fs: Full Screen
    canvas_fs_quad = np.array([0.0, 0.0, 0.0, 1.0, 1.0, 0.0, 1.0, 1.0]).astype('f4')
    vbo_fs = ctx.buffer(canvas_fs_quad.tobytes())
    vao_fs = ctx.simple_vertex_array(prog_for_depth, vbo_fs, 'in_vert')
    cbo_fs = ctx.renderbuffer(size, components=1, dtype='f4')
    fbo_fs = ctx.framebuffer(color_attachments=[cbo_fs])

    fbo_fs.use()
    depth_sampler = ctx.sampler(
        filter=(moderngl.LINEAR, moderngl.LINEAR),
        compare_func='',
    )
    depth_sampler.use(location=0)
    depth_texture.use(location=0)
    prog_for_depth['depth'].value = 0
    #
    vao_fs.render(moderngl.TRIANGLE_STRIP)

    #
    data = np.frombuffer(fbo_fs.read(components=1, dtype='f4'), dtype=np.dtype('f4')).reshape(size[::-1])
    print("From final fbo - some stuff should be 1s and 0.5 where the triangle lies. (like depth_texture)")
    print(data)


if __name__ == '__main__':
    main()
```

The expected result:
```bash
From depth texture - should all be 1
[[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]
From depth_texture - some stuff should be 1s and 0.5 where the triangle lies.
[[0.50000006 0.50000006 0.50000006 1.        ]
 [0.50000006 0.50000006 1.         1.        ]
 [0.50000006 1.         1.         1.        ]
 [1.         1.         1.         1.        ]]
From final fbo - some stuff should be 1s and 0.5 where the triangle lies. (like depth_texture)
[[0.50000006 0.50000006 0.50000006 1.        ]
 [0.50000006 0.50000006 1.         1.        ]
 [0.50000006 1.         1.         1.        ]
 [1.         1.         1.         1.        ]]
```

By the way, your project demosys-py look gorgeous ^^

**Comment by yoyonel** [comment]
My commit (on my fork repo) about this functionality: depth texture/buffer, RTT, samplers => https://github.com/yoyonel/ModernGL/commit/c20ecddaa8cc0434b4e0dc46123dd974df0b81e1
I'm using pytest to perform the unit test here: 
- https://github.com/yoyonel/ModernGL/blob/c20ecddaa8cc0434b4e0dc46123dd974df0b81e1/examples/conftest.py
- https://github.com/yoyonel/ModernGL/blob/c20ecddaa8cc0434b4e0dc46123dd974df0b81e1/examples/test_depth_samplers.py

**Comment by yoyonel** [comment]
Yes right, but my original issue was to retrieve and use depth informations from depth buffer/texture.
Using your advice (use a explicit sampler) resolve this issue.
Maybe we need to create a new issue about the potential problem about the "wrong" values in texture parameters.

**Comment by einarf** [comment]
I can look into it, no worries 😄

**Comment by einarf** [comment]
Fixed in #245. This was caused by incompleteness because the texture filtering was not set.

You actually don't need to use a sampler. Simply setting the filtering on the texture would also fix the issue:

```py
depth_texture.filter = filter=(moderngl.LINEAR, moderngl.LINEAR)
depth_texture.compare_func = ''
```

If the PR gets merge, then no need to set filter.

**Comment by einarf** [comment]
Verified that your initial example also now produce the correct result after the change
```
From depth texture - should all be 1
[[ 1.  1.  1.  1.]
 [ 1.  1.  1.  1.]
 [ 1.  1.  1.  1.]
 [ 1.  1.  1.  1.]]
From depth_texture - some stuff should be 1s and 0.5 where the triangle lies.
[[ 0.5  0.5  0.5  1. ]
 [ 0.5  0.5  1.   1. ]
 [ 0.5  1.   1.   1. ]
 [ 1.   1.   1.   1. ]]
From final fbo - some stuff should be 1s and 0.5 where the triangle lies. (like depth_texture)
[[ 0.50000006  0.50000006  0.50000006  1.        ]
 [ 0.50000006  0.50000006  1.          1.        ]
 [ 0.50000006  1.          1.          1.        ]
 [ 1.          1.          1.          1.        ]]
```

A final question. Is this depth only FBOs actually working for you?
```py
fbo_depth = ctx.framebuffer(depth_attachment=depth_texture)
```
I just get an error and wondering if this is some other bug...
```
mgl.Error: the depth_attachment have different sizes or samples
```

**Comment by yoyonel** [comment]
It's not a bug, this test not working with master branch (it run on a local branch).
This issue was an initial step before a pull request ^^
I've added slight fix/modification in source code (framebuffer.cpp)).
I will merge PR#245 and test it. If it's work, i will pr this test and my modifications.
Thx again !

**Comment by cprogrammer1994** [comment]
merged #245 :)
